### ISSUE TITLE
How do I use tsyringe for dependency injection?

### ISSUE THREAD

**Issue by emaxedon** [typeorm/typeorm#4790]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

<!--
    To answer those questions you need to put "x" inside the square brackets, for example:
    [x] `mysql`
    [ ] `postgres`

    Also, please format your code properly (by taking code blocks into ```ts .... ```)
!>

Currently there is a great example of how to use TypeORM with typedi. How would I go about using https://github.com/microsoft/tsyringe instead?

**Comment by emaxedon** [comment]
https://github.com/microsoft/tsyringe

**Comment by brunoMaurice** [comment]
This lib don't have documentation ?
What information from TypeOrm did you need ?

**Comment by FreakyBytes** [comment]
Found this issue via Google. I've got TypeORM working with TSyringe with the following snippet:

```ts
import { container, instanceCachingFactory } from 'tsyringe';
import { ConnectionManager, createConnections, useContainer } from 'typeorm';

// Register TypeORM's connection manager as singleton
// TypeORM does not register anything itself with TSyringe and therefore would get a new ConnectionManager every time
container.register<ConnectionManager>(ConnectionManager, { useFactory: instanceCachingFactory(() => new ConnectionManager()) });
// configure TypeORM to use a dependency injection container
useContainer(
  // wrapper because TypeORM expects `get` function from IoC container
  { get: someClass => container.resolve(someClass as any) },
);
const connections = await createConnections(options);
```

**Comment by fhenriques-meli** [comment]
@FreakyBytes I didn't understand how you use this code snippet. Can you explain me?

**Comment by FreakyBytes** [comment]
Sure thing @fhenriques-meli !

TypeORM provides a mechanism [to use an arbitrary Dependency Injection Container](https://github.com/typeorm/typeorm/blob/master/docs/faq.md#how-to-use-typeorm-with-a-dependency-injection-tool) via `useContainer`. However, TypeORM expects the DI container to have a `get` function - TSyringe has named it `resolve`, hence the thin wrapper in `useContainer`.

Another caveat is, that TypeORM does not register anything with the DI container. (it only expects a get function) TSyringe now has the caveat of rolling with unknown injection tokens. As long as they are classes, the DI container will just create an instance. Well, this is really bad in case of TypeORM's `ConnectionManager`, since this is the mechanism for TypeORM to check if a connection is available. Now, if TSyringe creates a new `ConnectionManager` every time TypeORM asks  - no connection will ever be available (to the rest of the application).
To counteract that, we manually register a [Singleton Factory](https://github.com/microsoft/tsyringe#instancecachingfactory) for it.

TL;DR: just call `container.register` with a singleton factory for the `ConnectionManager` and `useContainer` before creating any TypeORM connections :)

Hope this clears things up!

**Comment by trevor-gullstad** [comment]
@FreakyBytes what decorators do you use on your services and repositories?

**Comment by imnotjames** [comment]
For questions or other support, please check out the [community slack](https://join.slack.com/t/typeorm/shared_invite/zt-gej3gc00-hR~L~DqGUJ7qOpGy4SSq3g) or check [TypeORM's documentation page on other support avenues](https://typeorm.io/#/support) - cheers!
